avro-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaAvroDeserializer.java

```diff
@@ -252,4 +227,88 @@ private Schema getReaderSchema(Schema writerSchema) {
     }
     return readerSchema;
   }
+
+  class DeserializationContext {
+    private final String topic;
+    private final Boolean isKey;
+    private final ByteBuffer buffer;
+    private final int schemaId;
+
+    private Schema cachedSchemaFromRegistry = null;
+    private Schema cachedSchemaForDeserialize = null;
+
+    DeserializationContext(final String topic, final Boolean isKey, final byte[] payload) {
+      this.topic = topic;
+      this.isKey = isKey;
+      this.buffer = getByteBuffer(payload);
+      this.schemaId = buffer.getInt();
+    }
+
+    Schema getSchemaFromRegistry() {
+      if (cachedSchemaFromRegistry == null) {
+        try {
+          this.cachedSchemaFromRegistry = schemaRegistry.getById(schemaId);
+        } catch (RestClientException | IOException e) {
+          throw new SerializationException("Error retrieving Avro schema for id " + schemaId, e);
+        }
+      }
+      return cachedSchemaFromRegistry;
+    }
+
+    Schema getSchemaForDeserialize() {
+      if (cachedSchemaForDeserialize == null) {
+        try {
+          this.cachedSchemaForDeserialize = schemaForDeserialize(
+              schemaId, getSchemaFromRegistry(), getSubject(), isKey);
+        } catch (RestClientException | IOException e) {
+          throw new SerializationException("Error retrieving Avro schema for id " + schemaId, e);
+        }
+      }
+      return cachedSchemaForDeserialize;
+    }
+
+    String getSubject() {
+      return subjectName(topic, isKey, getSchemaFromRegistry());
+    }
+
+    String getTopic() {
+      return topic;
+    }
+
+    boolean isKey() {
+      return isKey;
+    }
+
+    int getSchemaId() {
+      return schemaId;
+    }
+
+    W read(Schema writerSchema) {
+      return read(writerSchema, null);
+    }
+
+    W read(Schema writerSchema, Schema readerSchema) {
+      DatumReader<W> reader = getDatumReader(writerSchema, readerSchema);
+      int length = buffer.limit() - 1 - idSize;
+      if (writerSchema.getType().equals(Schema.Type.BYTES)) {
+        byte[] bytes = new byte[length];
+        buffer.get(bytes, 0, length);
+        return (W) bytes;
+      } else {
+        int start = buffer.position() + buffer.arrayOffset();
+        try {
+          W result = reader.read(null, decoderFactory.binaryDecoder(buffer.array(),
+              start, length, null));
+          if (writerSchema.getType().equals(Schema.Type.STRING)) {
+            return (W) result.toString();
+          } else {
+            return result;
+          }
+        } catch (IOException e) {
+          throw new SerializationException("Error deserializing Avro message for id "
+              + schemaId, e);
```

Comment:
The old code was catching RuntimeException in case deserialization cause a NullPointerException, etc. Do we not want to keep that behavior?